src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing/DependencyInjectionValidation.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Fabric;
+using System.Linq;
+using System.Reflection;
+using Autofac;
+using Microsoft.DotNet.ServiceFabric.ServiceHost;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.Internal.DependencyInjection.Testing
+{
+    public static class DependencyInjectionValidation
+    {
+        private static readonly ImmutableList<Type> s_exemptTypes = ImmutableList.Create(
+            typeof(ServiceContext),
+            // IConfigure options is a strange type, and built into the options framework, no need to validation
+            typeof(IConfigureOptions<>),
+            // ILifetimeScope comes from Autofac, and we are only checking the ASP.NET half of this
+            typeof(ILifetimeScope),
+            typeof(MemoryCacheOptions)
+        );
+
+        private static readonly ImmutableList<string> s_exemptNamespaces = ImmutableList.Create(
+            "Microsoft.ApplicationInsights.AspNetCore"
+        );
+
+        public static bool IsDependencyResolutionCoherent(Action<ServiceCollection> register, bool includeServiceHost, out string errorMessage)
+        {
+            errorMessage = null;
+
+
+            var services = new ServiceCollection();
+            if (includeServiceHost)
+            {
+                Environment.SetEnvironmentVariable("ENVIRONMENT", "XUNIT");
+                ServiceHost.ConfigureDefaultServices(services);
+            }
+
+            register(services);
+
+            foreach (ServiceDescriptor service in services)
+            {
+                if (service.ImplementationType == null)
+                {
+                    continue;
+                }
+
+                if (IsExemptType(service.ImplementationType) || IsExemptType(service.ServiceType))
+                {
+                    continue;
+                }
+
+                bool foundConstructable = false;
+                ConstructorInfo[] constructors = service.ImplementationType
+                    .GetConstructors(BindingFlags.Public | BindingFlags.Instance)
+                    .OrderBy(c => c.GetParameters().Length)
+                    .ToArray();
+
+                if (constructors.Length == 0)
+                {
+                    // zero constructor things are implicitly constructable
+                    continue;
+                }
+
+                foreach (ConstructorInfo ctor in constructors)
+                {
+                    bool ctorWorks = true;
```

Comment:
I think this loop logic isn't exactly what Autofac/ServiceProvider do. They pick the longest constructor and use that, they don't attempt to use shorter constructors if they can't resolve services for the longer constructor.

I don't know how many of our services have multiple constructors, but this logic would always say "its fine" on a type with a constructor with no parameters.

Reply:
It seems to choose the longest that it can find all the types for.  We have one type in Helix that takes an optional "IHelixDataProvider", and it seems to resolve it both when I register the type and not.

Reply:
Thats interesting.... I guess that this is correct then.

Reply:
Yeah, I tested it with this: https://gist.github.com/ChadNedzlek/0ebc2d60805e2623d0936a8a02634176, and it output:
```
Thing with Secret
Thing
```

Implying it chose which constructor to resolve based on available registrations.